File: java/src/main/java/cn/az/code/year2020/jun/CoinChange2.java
package cn.az.code.year2020.jun;

public class CoinChange2 {

    public static void main(String[] args) {
        CoinChange2 cc = new CoinChange2();
        int[] coins = new int[]{1,2,5};
        System.out.println(cc.change(12,coins));
    }

    public int change(int amount, int[] coins) {
        int[] dp = new int[amount + 1];
        dp[0] = 1;
        for (int coin : coins) {
            for (int i = coin; i <= amount; i++) {
                dp[i] += dp[i-coin];
            }
        }
        return dp[amount];
    }
}

File: code/python/2020/june/coin-change2.py
from typing import List


def change(amount: int, coins: List[int]) -> int:
    dp = [0 for _ in range(amount + 1)]
    dp[0] = 1
    for c in coins:
        for i in range(c, amount + 1):
            dp[i] += dp[i - c]
    return dp[amount]
